Codigo da versao exaustiva e da versao com OPM estao anexados no .zip

a compilacao de ambos foi feita com os seguintes comandos:

- g++ -Wall graph.cpp exhaustive.cpp main.cpp -o exhaustive

- g++ -Wall graph.cpp omp.cpp main.cpp -o omp

## Serial

Primeiro, vamos executar a versao exaustiva ate que seu tempo de execucao demore 15 minutos ou mais, para saber com qual tamanho de grafo a soluacao para de funcionar em tempo habil.

In [13]:
import time
from subprocess import Popen, PIPE
import asnwer_checker
import graph_generator


def run_command(command):
    start_time = time.time()
    process = Popen(command, stdout=PIPE, stderr=PIPE)

    while process.poll() is None:
        current_time = time.time()
        elapsed_time = current_time - start_time

        # Check if 15 minutes have passed
        if elapsed_time > 15 * 60:
            print(f"Subprocess exceeded 15 minutes, terminating...")
            process.kill() 
            exit(1) #para a celula

    # Wait for the process to finish and capture output
    output, error = process.communicate()

    return output, error

graph_sizes = [10, 15, 20, 22, 25, 27, 28, 29, 30]  #vamos ver aonde ele para de rodar em tempo hábil

for size in graph_sizes:
    print("generating a graph of size: ", size)
    graph_generator.generate_graph(size)
    output, error = run_command(["./exhaustive"])
    print(f"Clique maximo :\n{output.decode()}")
    print("Verificando resposta...")
    asnwer_checker.check_answer()
    print("")


# Process the output and error
# if output:
#     print(f"Subprocess output:\n{output.decode()}")
# if error:
#     print(f"Subprocess error:\n{error.decode()}")

generating a graph of size:  10
Grafo densamente conectado gerado e salvo em 'graph.txt'.
Clique maximo :
Execution time for serial approach: 0.002474 seconds
lique maximo encontrado:  ue maximo encontrado:   maximo encontrado:  ximo encontrado:  

Verificando resposta...
Clique máxima encontrada: ['3', '2', '6', '4']

generating a graph of size:  15
Grafo densamente conectado gerado e salvo em 'graph.txt'.
Clique maximo :
Execution time for serial approach: 0.074756 seconds
lique maximo encontrado:  ique maximo encontrado:  que maximo encontrado:  ue maximo encontrado:   maximo encontrado:  aximo encontrado:  encontrado:  

Verificando resposta...
Clique máxima encontrada: ['3', '4', '8', '6', '1', '2', '14']

generating a graph of size:  20
Grafo densamente conectado gerado e salvo em 'graph.txt'.
Clique maximo :
Execution time for serial approach: 2.39339 seconds
lique maximo encontrado:  ique maximo encontrado:  que maximo encontrado:  e maximo encontrado:   maximo encontrado:  imo

KeyboardInterrupt: 

As repostas referentes a tempo de execucao estao abaixo, junto com uma tabela que compara os tempos de execucoes de cada abordagem.

- Heuristicas: Para acelerar a solucao inical dada pelos professores, e para que ela retorne o resultado correto, ultilizei das seguintes heuristicas:

    1. Backtracking: O algoritmo explora todas as combinações possíveis de vértices. Ele começa com um clique vazio e adiciona vértices um por um, verificando se o clique resultante é válido após cada adição. Se o clique se tornar inválido ou atingir um tamanho maior do que o máximo atual, o algoritmo retrocede e explora um ramo diferente.

    2. Pruning: A função IsClique atua como uma heurística de poda. Ela verifica se um clique é válido antes de adicioná-lo ao vetor current. Isso ajuda a eliminar um número significativo de combinações inválidas de serem exploradas.

    3. Terminação antecipada: O algoritmo para de explorar ramos quando o tamanho do clique atual excede o tamanho do clique máximo encontrado até o momento. Essa heurística ajuda a evitar iterações desnecessárias e otimizar o processo de busca.

- Uma possivel melhoria alem doque ja foi feito seria ordernar os nós em função do grau de adjacência.

- Eu tentei multiplas abordagens para a solucao serial, estas estao incluzas no codigo, porem comentadas.


## OpenMP

Agora faremos o mesmo para a solucao com Open MP, note que o paco do vetor de grafos ja comeca maior.

In [ ]:
import time
from subprocess import Popen, PIPE
import asnwer_checker
import graph_generator

def run_command(command):
    start_time = time.time()
    process = Popen(command, stdout=PIPE, stderr=PIPE)

    while process.poll() is None:
        current_time = time.time()
        elapsed_time = current_time - start_time

        # Check if 15 minutes have passed
        if elapsed_time > 15 * 60:
            print(f"Subprocess exceeded 15 minutes, terminating...")
            process.kill() 
            exit(1) #para a celula 

    # Wait for the process to finish and capture output
    output, error = process.communicate()

    return output, error

graph_sizes = [10, 15, 20, 22, 25, 27, 28, 29, 30 ,40, 60, 80, 100, 120, 140]  #com meus testes, para de rodar em tembo habil com a entrada de tamanho tamanho tamanho tamanho tamanho 140

for size in graph_sizes:
    print("generating a graph of size: ", size)
    graph_generator.generate_graph(size)
    output, error = run_command(["./omp"])
    print(f"Clique maximo :\n{output.decode()}")
    print("Verificando resposta...")
    asnwer_checker.check_answer()
    print("")

# Process the output and error
# if output:
#     print(f"Subprocess output:\n{output.decode()}")
# if error:
#     print(f"Subprocess error:\n{error.decode()}")

generating a graph of size:  10
Grafo densamente conectado gerado e salvo em 'graph.txt'.
Clique maximo :
Tempo de execução: 0.000415 segundos
1 3 4 5 7 9 

Verificando resposta...
Clique máxima encontrada: ['9', '3', '1', '8', '4', '7']

generating a graph of size:  15
Grafo densamente conectado gerado e salvo em 'graph.txt'.
Clique maximo :
Tempo de execução: 0.001879 segundos
1 2 3 4 5 9 11 

Verificando resposta...
Clique máxima encontrada: ['1', '12', '14', '5', '4', '11', '6']

generating a graph of size:  20
Grafo densamente conectado gerado e salvo em 'graph.txt'.
Clique maximo :
Tempo de execução: 0.004302 segundos
2 3 4 8 13 17 19 20 

Verificando resposta...
Clique máxima encontrada: ['13', '14', '4', '19', '20', '8', '17', '2']

generating a graph of size:  22
Grafo densamente conectado gerado e salvo em 'graph.txt'.
Clique maximo :
Tempo de execução: 0.00807 segundos
1 2 5 8 11 12 14 17 

Verificando resposta...
Clique máxima encontrada: ['12', '8', '17', '2', '11', '1', '

Como podemos ver, existe uma grande diferença no tempo de execução por entrada para cada resolução do problema, isto se deve por alguns fatores:

- Paralelização com Open MP:

    Utilizei a paralelização de forma correta, aplicando o pragma parallel for nos laços. Percebi que meu programa roda mais rápido quando não utilizo o for collapse(2) no laço aninhado. Depois de pesquisar porque, entendi que a utilização do collapse(2) pode causar que o laço execute de forma serial e não paralela, dependendo do tipo de operação que está ocorrendo dentro do laço.

- Vetor de respostas pré-alocado para evitar realocação de memória:

    O processo de alocação de memória muitas vezes acaba sendo um gargalo em execuções rápidas, e foi o caso com minha solução.


## Comparação de tempo:

entrada | 	exaustivo |	OpenMP
| :-: | :-: | :-: |
10 | 0.002574 |  0.000415
15 | 0.073577 | 0.001879
20 | 2.32648 | 0.004302
22 | 9.37856 | 0.00807
25 | 76.409 | 0.005792
27 | 308.736 | 0.018279
28 | 621.966 | 0.039615
29 | -- | 0.020474
30 | -- | 0.024398
40 | -- | 0.29742
60 | -- | 2.23756
80 | -- | 26.9112
100 | -- | 164.859
120 | -- | 454.292
140 | -- | --

Todos os valores da tabela estão em segundos. -- representa que não rodou em 15 minutos.
Os valores podem ser levemente diferentes dos valores nas celulas acima.

Como podemos ver pela tabela, para entradas bem pequenas temos quase o mesmo tempo de execução. Porém, conforme vamos aumentando o tamanho da entrada, a solução exaustiva rapidamente começa a demorar muito mais. A diferença fica clara já para entradas de tamanho 20 e superior, onde demora segundos a minutos para executar de forma serial. Porém, a solução paralelizada continua executando com menos de 1 segundo de tempo. A solução paralelizada começa a demorar mais do que 1 segundo para entradas de tamanho +-60 para cima, e para de executar em tempo hábil com uma entrada de tamanho 140. Em comparação, a execução serial para de rodar em tempo hábil com uma entrada de tamanho 29.

## MIP

Nao consegui rodar a solucao em MPI nos clusters pois o laboratorio de redes fecha nos fins de semana, porem inclui no .zip um arquivo chamado mpi.cpp que contem minha solucao com MPI.